### PR TITLE
fix(memory): Protect blocks from being marked as reclaimable multiple times.

### PR DESCRIPTION
**Pull request by broneill**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :**
A background task can call BlockManager.markBucketedBlocksReclaimable, which can then steal blocks which are owned by BlockMemFactory instances. Because the BlockMemFactory isn't notified, it thinks it can still access the current block and mark it reclaimable again later. This makes it possible for the block to be shared, leading to corruption.

**New behavior :**
If there is a block owner assigned, it's consulted first. It can reject the attempt to mark its current block as reclaimable, if it's still being written to. If the attempt is accepted, then all blocks tracked by mem factory are marked as reclaimable, and all references to them are discarded.

### CODE REVIEW

File: memory/src/main/scala/filodb.memory/MemFactory.scala [hunk 1]
```diff
@@ -212,18 +219,51 @@ class BlockMemFactory(blockStore: BlockManager,
...
+
+  private def requestBlock() = blockStore.requestBlock(bucketTime, optionSelf).get
 
   // tracks blocks that should share metadata
   private val metadataSpan: ListBuffer[Block] = ListBuffer[Block]()
 
+  // Last time this factory was used for allocation.
+  private var lastUsedNanos = now
+
+  private def now: Long = System.nanoTime()
```
Comment (velvia) [inline memory/src/main/scala/filodb.memory/MemFactory.scala]: How fast is nanoTime call?  
The other concern is that nanoTime can potentially wrap around, can't it?

Comment (broneill) [reply memory/src/main/scala/filodb.memory/MemFactory.scala]: It's plenty fast. Consider also that the block allocation is relatively infrequent, so not something to be concerned with. The nanotime value can wrap around, which is why modulo comparison (subtraction) is required.

File: core/src/main/scala/filodb.core/memstore/DemandPagedChunkStore.scala [hunk 2]
```diff
@@ -73,21 +73,23 @@ extends RawToPartitionMaker with StrictLogging {
       tsShard.shardStats.numChunksPagedIn.increment(rawPartition.chunkSets.size)
       // One chunkset at a time, load them into offheap and populate the partition
       rawPartition.chunkSets.foreach { case RawChunkSet(infoBytes, rawVectors) =>
-        val memFactory = getMemFactory(timeBucketForChunkSet(infoBytes))
-        val chunkID = ChunkSetInfo.getChunkID(infoBytes)
+        if (!rawVectors.isEmpty) {
```
Comment (velvia) [inline core/src/main/scala/filodb.core/memstore/DemandPagedChunkStore.scala]: Please add a comment detailing the bug about metadata allocation error that might possibly happen that this fixes

**Comment by broneill** [comment]
Also adding code to "Return the last metaAddr from endMetaSpan which actually has room" doesn't guarantee that the returned address is non-zero. The first might have been full as well. The newly added comment explains my motivation for just a simple fix.

**Pull request merged**